### PR TITLE
Layout features

### PR DESCRIPTION
**Pull request by RoelN**
TODO

- [x] ssXX and cvXX
- [x] split fixed and on/off
- [x] Split GSUB/GPOS

ON HOLD:

The following stuff is kept back for now, because we need to check if this is the way to go, and if/how it's possible to do so with Font.js.

- [ ] group features by langsys[1]
- [ ] Link feature to optional name from name table (to be implemented in Font.js)
- [ ] Return characters per feature (how? :D)

[1] Currently just all features in the font are shown, regardless of which script/langsys they belong too. Is that good enough?? Currently investigating!

### CODE REVIEW

File: src/fondue/Fondue.js [hunk 1]
```diff
@@ -20,20 +20,37 @@ export default class Fondue {
 	}
 
 	get isColor() {
-		return this.colorFormats.length >= 1;
+		return this.colorFormats.length > 0;
+	}
+
+	get hasFeatures() {
+		return (
```
Comment (pascalw) [inline src/fondue/Fondue.js:27]: Can there be other kind of features besides `fixed` or `optional`?

File: src/fondue/Fondue.js [hunk 2]
```diff
@@ -20,20 +20,37 @@ export default class Fondue {
...
 	get languageSystems() {
 		const getLangs = (table) => {
-			return table
-				.getSupportedScripts()
-				.reduce((acc, script) => {
-					const scriptTable = table.getScriptTable(script);
-					return acc.concat(table.getSupportedLangSys(scriptTable));
-				}, [])
-				.map((lang) => lang.trim());
+			if (table) {
```
Comment (pascalw) [inline src/fondue/Fondue.js:41]: It seems like this check could be moved to the outer function, as guard clause? I.e.:

```js
const gsubTable = this._font.opentype.tables.GSUB;
if(! gsubTable) return [];

const gsubLangs = getLangs(gsubTable);
...

Makes the nested function more readable and in general I think it's good to abort early.
```

File: src/fondue/Fondue.js [hunk 3]
```diff
@@ -209,37 +226,52 @@ export default class Fondue {
...
+							features["optional"][feature] =
+								featureMapping[featureIndex];
+						}
+					}
+				});
+			}
+		};
+
+		getFeats(this._raw("GSUB"));
+		getFeats(this._raw("GPOS"));
```
Comment (pascalw) [inline src/fondue/Fondue.js:272]: Hmm this code really makes it seem like it's doing nothing, but it's modifying a variable in scope `const features`.
Could you change this so it's not mutating this variable?

It seems that the code doesn't have to distinguish between GSUB and GPOS, so perhaps you could something like this:

```js
const getRawFeatures = table => tabe.featureList.featureRecords.map(
    (record) => record.featureTag
);

const rawFeatures = [
   ...getRawFeatures(this._raw("GSUB")),
   ...getRawFeatures(this._raw("GPOS")),
];

return rawFeatures.reduce((features, rawFeature) => {
  // mapping
  features["fixed"][rawFeature] = ...;
}, {});
```

File: src/tools/features/layout-features.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // OpenType Layout Features
 //
 // Based on "Enabling Typography: towards a general model of
-// OpenType Layout" by John Hudson, Tiro Typoworks
```
Comment (pascalw) [inline src/tools/features/layout-features.js:4]: 😅

Comment (RoelN) [reply src/tools/features/layout-features.js:4]: The perfect typo 😆

File: src/fondue/Fondue.js (hunk 2 above)
Comment (RoelN) [reply src/fondue/Fondue.js:38]: With the new code I'm still catching it in the code that gets the raw features. Please see if this is okay!

File: src/fondue/Fondue.js (hunk 1 above)
Comment (RoelN) [reply src/fondue/Fondue.js]: No, only these two types. But with the refactor, this is no longer an issue anyway :-)

File: src/tools/css/get-css.js [hunk 5]
```diff
@@ -105,8 +109,11 @@ const getCSS = (fondue) => {
 		let cssvardecs = "";
 
 		for (const feature of features) {
-			const indexedFeature = getFeatureIndex(feature);
-			const defaultState = layoutFeatures[indexedFeature].state;
+			const featureIndex = getFeatureIndex(feature);
```
Comment (pascalw) [inline src/tools/css/get-css.js:112]: Not for this PR but it's probably worthwhile to add some tests for the CSS output. It's a substantial amount of code that can easily break if something in the engine changes.

**Review by pascalw — APPROVED** [review]

File: src/tools/css/get-css.js (hunk 5 above)
Comment (RoelN) [reply src/tools/css/get-css.js:112]: That is true. CSS will be handled shortly, so let's make it more robust in a future PR!

**Pull request merged**